### PR TITLE
Adding some tensor operations around softmaxes and masks

### PR DESCRIPTION
**Pull request by matt-gardner**
I originally had these in the BiDAF PR, but they are independent and can be reviewed / merged on their own.  Makes for a simpler and easier PR to review to have them separate.

### CODE REVIEW

File: allennlp/common/tensor.py [hunk 1]
```diff
@@ -123,3 +123,68 @@ def masked_softmax(vector, mask):
...
+    Takes a matrix of vectors and a set of weights over the rows in the matrix (which we call an
+    "attention" vector), and returns a weighted sum of the rows in the matrix.  This is the typical
+    computation performed after an attention mechanism.
+
+    Note that while we call this a "matrix" of vectors and an attention "vector", we also handle
+    higher-order tensors.  We always sum over the second-to-last dimension of the "matrix", and we
+    assume that all dimensions in the "matrix" prior to the last dimension are matched in the
+    "vector".  Non-matched dimensions in the "vector" must be `directly after the batch dimension`.
+
+    For example, say I have a "matrix" with dimensions ``(batch_size, num_queries, num_words,
```
Comment (DeNeutoy) [inline allennlp/common/tensor.py]: This section is unnecessarily verbose - I think cutting this down to make it a more concrete example would be helpful. 
```
For example, say I have a "matrix" with dimensions (batch_size, num_queries, num_words, embedding_dim). The attention "vector" then must have at least those dimensions, and could have more. Both:

- ``(batch_size, num_queries, num_words)`` (distribution over words for each query)
- ``(batch_size, num_documents, num_queries, num_words)`` (distribution over words in a query for each document)

are valid input ``vectors``, producing tensors of shape: ``(batch_size, num_queries, embedding_dim)`` and ``(batch_size, num_documents, num_queries, embedding_dim)`` respectively.

```

File: allennlp/common/tensor.py [hunk 2]
```diff
@@ -123,3 +123,68 @@ def masked_softmax(vector, mask):
     else:
         # There is no mask, so we use the provided ``torch.nn.functional.softmax`` function.
         return torch.nn.functional.softmax(vector)
+
+
+def get_text_field_mask(text_field_tensors: Dict[str, torch.Tensor]) -> torch.Tensor:
```
Comment (DeNeutoy) [inline allennlp/common/tensor.py]: I don't exactly see the need for this to be a function - we are making a big assumption about the name of the token field. This also applies to the mask function I wrote above, btw. What higher order functionality were you imagining?

Comment (matt-gardner) [inline allennlp/common/tensor.py]: I wanted it to be a function so that the model code doesn't have to think about the content of the text field arrays at all.  For higher-order tensors, I was thinking of things like memory network inputs, sets of paragraphs for TriviaQA, or other places you have a `ListField[TextField]`.  To handle this and remove the magical "tokens" key, I think the right thing to do is just find the entry in the dictionary that has the minimum number of dimensions, and use that.

File: allennlp/common/tensor.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/common/tensor.py]: Yeah, that's better.
